test/framework/template.go

```diff
@@ -0,0 +1,89 @@
+package framework
+
+import (
+	"context"
+	"os"
+	"testing"
+
+	anywherev1 "github.com/aws/eks-anywhere/pkg/api/v1alpha1"
+	releasev1 "github.com/aws/eks-anywhere/release/api/v1alpha1"
+)
+
+// ProviderTemplateNameGenerator is an interface for getting template name for each provider.
+type ProviderTemplateNameGenerator interface {
+	EnvVarForTemplate(osFamilyStr, eksDName string) string
+	DefaultNameForTemplate(osFamilyStr, eksDName string) string
+	SearchTemplate(ctx context.Context, template string) (string, error)
+	DefaultEnvVarForTemplate(osFamilyStr string, kubeVersion anywherev1.KubernetesVersion) string
+}
+
+// TemplateCache is a map of template name.
+type TemplateCache struct {
+	cache map[string]string
+}
+
+// TemplateForRelease tries to find a suitable template for a particular eks-a release, k8s version and OS family.
+// It follows these steps:
+//
+// 1. Look for explicit configuration through an env var: "T_{provider}_TEMPLATE_{osFamily}_{eks-d version}"
+// eg. T_CLOUDSTACK_TEMPLATE_REDHAT_KUBERNETES_1_23_EKS_22, T_VSPHERE_TEMPLATE_REDHAT_KUBERNETES_1_23_EKS_22
+// This should be used for explicit configuration, mostly in local development for overrides.
+//
+// 2. If not present, look for a template if the default templates: "{eks-d version}-{osFamily}"
+// eg. kubernetes-1-23-eks-22-redhat (CloudStack),  /SDDC-Datacenter/vm/Templates/kubernetes-1-23-eks-22-redhat (vSphere)
+// This is what should be used most of the time in CI, the explicit configuration is not present but the right template has already been
+// imported to cloudstack.
+//
+// 3. If the template doesn't exist, default to the value of the default template env vars: eg. "T_CLOUDSTACK_TEMPLATE_REDHAT_1_23".
+// This is a catch all condition. Mostly for edge cases where the bundle has been updated with a new eks-d version, but the
+// the new template hasn't been imported yet. It also preserves backwards compatibility.
+func (tc *TemplateCache) TemplateForRelease(t *testing.T, generator ProviderTemplateNameGenerator, osFamily anywherev1.OSFamily, release *releasev1.EksARelease, kubeVersion anywherev1.KubernetesVersion) string {
```

Comment:
I think we can have `generator` as a field in the `TemplateCache` to avoid having to pass it every time